### ISSUE TITLE
[Bug]: InvalidOperationException in NuGet.VisualStudio.VsPackageInstaller.InstallPackage method before building the solution

### ISSUE THREAD

**Issue by TomW-Skyline** [NuGet/Home#11910]
### NuGet Product Used

NuGet SDK

### Product Version

Visual Studio 17.1.2

### Impact

I'm unable to use this version

### Repro Steps & Context

When adding a reference using the Nuget API (NuGet.VisualStudio.VsPackageInstaller.InstallPackage method) from within a Visual Studio extension the exception below occurs.
This only happens when we haven't built the solution yet. When we first build the solution, and then add the Nuget to a project it works.

System.InvalidOperationException: Failed to add reference. The package 'Newtonsoft.Json' tried to add a framework reference to 'System' which was not found in the GAC. This is possibly a bug in the package. Please contact the package owners for assistance. ---> System.Runtime.InteropServices.COMException: Reference unavailable. at VSLangProj.References.Add(String bstrPath)
at NuGet.PackageManagement.VisualStudio.VsMSBuildProjectSystem.<AddFrameworkReferenceAsync>d__78.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
--- End of inner exception stack trace ---
at NuGet.PackageManagement.VisualStudio.VsMSBuildProjectSystem.<AddFrameworkReferenceAsync>d__78.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at NuGet.ProjectManagement.MSBuildNuGetProject.<InstallPackageAsync>d__36.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at NuGet.PackageManagement.NuGetPackageManager.<ExecuteInstallAsync>d__89.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at NuGet.PackageManagement.NuGetPackageManager.<ExecuteNuGetProjectActionsAsync>d__79.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at NuGet.PackageManagement.NuGetPackageManager.<ExecuteNuGetProjectActionsAsync>d__79.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at NuGet.PackageManagement.NuGetPackageManager.<ExecuteNuGetProjectActionsAsync>d__79.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at NuGet.PackageManagement.NuGetPackageManager.<InstallPackageAsync>d__48.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at NuGet.PackageManagement.NuGetPackageManager.<InstallPackageAsync>d__47.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at NuGet.VisualStudio.VsPackageInstaller.<InstallInternalCoreAsync>d__25.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at NuGet.VisualStudio.VsPackageInstaller.<InstallInternalAsync>d__24.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at NuGet.VisualStudio.VsPackageInstaller.<InstallInternalAsync>d__24.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()
at NuGet.VisualStudio.VsPackageInstaller.InstallPackage(String source, Project project, String packageId, String version, Boolean ignoreDependencies)

**Comment by kartheekp-ms** [comment]
Can you please provide us the steps to reproduce this issue? I guess https://docs.microsoft.com/nuget/visual-studio-extensibility/nuget-api-in-visual-studio#nugetvisualstudio we should be coding against interfaces not the `VsPackageInstaller` implementation itself. 

https://docs.microsoft.com/nuget/visual-studio-extensibility/nuget-api-in-visual-studio#using-nuget-services doc has some help on how to use NuGet services in Visual Studio. Please take a look and let us know your feedback.

**Comment by TomW-Skyline** [comment]
Sorry my mistake, I had copy pasted that method name from the stacktrace.
We are using the interfaces as described in the links that you've mentioned above.
A collegue will try to create a solution that reproduces the problem.

**Comment by PedroDebevere** [comment]
Hi, I created a basic VS extension to reproduce it. The extension can be found on [https://github.com/PedroDebevere/ExampleExtension](https://github.com/PedroDebevere/ExampleExtension). The extension provides an additional menu item under the Tools menu named "Test Install NuGet". Clicking this item will try to install the FluentAssersion NuGet package (version 5.10.3) to all projects in the solution. Once completed, a message box should be shown with the projects where it successfully installed it or an exception, if any.

An example solution that results in an exception during installation of this package can be found here: [https://github.com/PedroDebevere/ExampleSolution](https://github.com/PedroDebevere/ExampleSolution). When opening the solution and then pressing the "Test Install NuGet" from the extension, it should show an exception stating that it tried to add a framework reference to 'System.Configuration' which was not found in the GAC.

When trying to install the NuGet package via the package manager in VS itself, it works.

Note: When opening the solution and first performing a build before doing the install via the extension it should also succeed. Only when opening the solution and then performing the installation of the NuGet package via the extension that it would fail (in the Solution Explorer of VS you should also see a exclamation mark next to the System reference in the test project).

**Comment by erdembayar** [comment]
> Hi, I created a basic VS extension to reproduce it. The extension can be found on https://github.com/PedroDebevere/ExampleExtension. The extension provides an additional menu item under the Tools menu named "Test Install NuGet". Clicking this item will try to install the FluentAssersion NuGet package (version 5.10.3) to all projects in the solution. Once completed, a message box should be shown with the projects where it successfully installed it or an exception, if any.
> 
> An example solution that results in an exception during installation of this package can be found here: https://github.com/PedroDebevere/ExampleSolution. When opening the solution and then pressing the "Test Install NuGet" from the extension, it should show an exception stating that it tried to add a framework reference to 'System.Configuration' which was not found in the GAC.
> 

It looks you're testing with legacy type of project. Did you try with packagereference type project? Try and see what happens. You can create one easily with `dotnet new classlib mylib` command. 

> When trying to install the NuGet package via the package manager in VS itself, it works.
> 
> Note: When opening the solution and first performing a build before doing the install via the extension it should also succeed. Only when opening the solution and then performing the installation of the NuGet package via the extension that it would fail (in the Solution Explorer of VS you should also see a exclamation mark next to the System reference in the test project).

I'm not expert on VS extension programming. But few things I can observe from here is, in general VS solution building creates obj>>dgspec.json and asset.json files. Most likely your `install via the extension` operation is expecting `dependency graph` already generated by time it runs, maybe taking it as input, if it doesn't exist then it tries to fallback for GAC where .netframework framework installs assemblies. Please try check for dg spec input in your code and also test with .Net core projects as I suggested.

**Comment by nkolev92** [comment]
I can repro this issue. Repro steps:

1. Download ExampleExtension and run it with VS experimental instance.
2. Load the ExampleSOlution.
3. CLikc the command
4. Observe the failrue.

I also get the same behavior if I build before running the command the first time.

Getting a failure and then running build and attempting the install again doesn't seem to fix it.

When was the first time you observed this @PedroDebevere Which versions? Did this ever work?

Building first seems to initialize certain caches/look ups in the project system that makes it work. What's weird is that the PM UI always works.

**Comment by TomW-Skyline** [comment]
@nkolev92 thank you for the investigation, we are glad you could also reproduce it. We are not aware of a version in which it worked, so it could be it never worked.

**Comment by zivkan** [comment]
Firstly, I want to apologise for taking so long to look at this. BIG thank you to @PedroDebevere for creating and sharing the repro. I first tried to repro myself using our team's extenion to test our VS APIs, and a test project I created myself. It didn't repro. But the examples still repro in VS 17.8.

Ultimately, it's the VS project system throwing an exception (well, returning a COM error), so I created this Developer Community bug report, so it can be routed to the correct team: https://developercommunity.visualstudio.com/t/VSLangProjReferencesAdd-throws-before/10450393

I'm going to close this one as a duplicate of that one, since it's out of the NuGet team's direct control to fix this, but I will follow up with the project system team as needed.
